src/persist-client/src/usage.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Introspection of storage utilization by persist
+
+use mz_ore::metrics::MetricsRegistry;
+use std::sync::Arc;
+
+use mz_persist::cfg::BlobConfig;
+use mz_persist::location::{Blob, ExternalError};
+
+use crate::{retry_external, Metrics, ShardId};
+
+/// Provides access to storage usage metrics for a specific Blob
+#[derive(Debug)]
+pub struct StorageUsageClient {
+    blob: Arc<dyn Blob + Send + Sync>,
+}
+
+impl StorageUsageClient {
+    /// Creates a new StorageUsageClient pointed to a specific Blob
+    pub async fn open(blob_uri: &str) -> Result<StorageUsageClient, ExternalError> {
+        let blob = BlobConfig::try_from(blob_uri).await?;
+
+        // TODO: what do we want to do about the metrics here?
+        let metrics_registry = MetricsRegistry::new();
+        let metrics = Metrics::new(&metrics_registry);
+        let blob =
+            retry_external(&metrics.retries.external.blob_open, || blob.clone().open()).await;
+
+        Ok(StorageUsageClient { blob })
+    }
+
+    /// Returns the size (in bytes) of a shard's data stored within this Blob
+    pub async fn shard_size(&self, shard_id: &ShardId) -> Result<u64, ExternalError> {
+        let mut total_size = 0;
+        self.blob
+            .list_keys_and_metadata(Some(&shard_id.to_string()), &mut |metadata| {
```

Comment:
can we add a method to BlobPaths that gives a prefix back for a ShardId (I'm hoping to push all the key logic into a single place)

Reply:
not sure I follow -- is this to relabel the `to_string()` call? where would it get used/passed around?

Reply:
added `BlobKeyPrefix` as a separate enum type that's available at the `persist-client` and above level, and made the `StorageUsageClient`'s function a more generic `size` calculation over a given prefix, rather than only working for a specific `ShardId`. might be a little overdoing it, but it made everything a bit safer and easier to test 😇